src/server/service/kv.rs

```diff
@@ -82,29 +102,9 @@ impl<T: RaftStoreRouter + 'static, E: Engine> Service<T, E> {
 }
 
 impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E> {
-    fn kv_get(&self, ctx: RpcContext, mut req: GetRequest, sink: UnarySink<GetResponse>) {
+    fn kv_get(&self, ctx: RpcContext, req: GetRequest, sink: UnarySink<GetResponse>) {
         let timer = GRPC_MSG_HISTOGRAM_VEC.kv_get.start_coarse_timer();
-
-        let future = self
-            .storage
-            .async_get(
-                req.take_context(),
-                Key::from_raw(req.get_key()),
-                req.get_version(),
-            )
-            .then(|v| {
-                let mut resp = GetResponse::new();
-                if let Some(err) = extract_region_error(&v) {
-                    resp.set_region_error(err);
-                } else {
-                    match v {
-                        Ok(Some(val)) => resp.set_value(val),
-                        Ok(None) => (),
-                        Err(e) => resp.set_error(extract_key_error(&e)),
-                    }
-                }
-                Ok(resp)
-            })
+        let future = future_get(&self.storage, req)
```

Comment:
Seems that the things left behind looks nearly identical for every interface. Maybe we can just extract an function?

Reply:
Metrics and responses are different, so even if we can extract a function, we probably need pass closures instead of variables to it. So let's keep it?